Modules/itertoolsmodule.c

```diff
@@ -355,8 +363,30 @@ pairwise_next(pairwiseobject *po)
         Py_DECREF(old);
         return NULL;
     }
-    /* Future optimization: Reuse the result tuple as we do in enumerate() */
-    result = PyTuple_Pack(2, old, new);
+
+    result = po->result;
+    if (Py_REFCNT(result) == 1) {
+        Py_INCREF(result);
+        PyObject *last_old = PyTuple_GET_ITEM(result, 0);
+        PyObject *last_new = PyTuple_GET_ITEM(result, 1);
+        PyTuple_SET_ITEM(result, 0, Py_NewRef(old));
+        PyTuple_SET_ITEM(result, 1, Py_NewRef(new));
+        Py_DECREF(last_old);
+        Py_DECREF(last_new);
+        // bpo-42536: The GC may have untracked this result tuple. Since we're
+        // recycling it, make sure it's tracked again:
+        if (!_PyObject_GC_IS_TRACKED(result)) {
+            _PyObject_GC_TRACK(result);
+        }
+    }
+    else {
+        result = PyTuple_New(2);
```

Comment:
Should we set `po->result` to `result` here?

Reply:
I tried this diff:
```python
diff --git a/Modules/itertoolsmodule.c b/Modules/itertoolsmodule.c
index 21ce3ecfad..9c3674d7d2 100644
--- a/Modules/itertoolsmodule.c
+++ b/Modules/itertoolsmodule.c
@@ -384,6 +384,7 @@ pairwise_next(pairwiseobject *po)
         if (result != NULL) {
             PyTuple_SET_ITEM(result, 0, Py_NewRef(old));
             PyTuple_SET_ITEM(result, 1, Py_NewRef(new));
+            Py_SETREF(po->result, Py_NewRef(result));
         }
     }
```
With results:
```
b1(1000)         min=28.4ms mean=29.0ms ± 0.2ms (50 repeats, 1000 loops)
b2(1000)         min=16.9ms mean=17.2ms ± 0.2ms (50 repeats, 1000 loops)
b3(1000)         min=26.0ms mean=26.6ms ± 0.4ms (50 repeats, 1000 loops)
```

It removed any benefit in the `for x in pairwise(range(n)):` (b1) microbenchmark, which makes sense. It seems at best neutral on b2 and b3.

Note that I copied the implementation from enumobject, which also does not update.

Reply:
Thank you for your investigations. There are cases in which updating `po->result` may be beneficial, e.g:
```py
it = pairwise(...)
first = next(it)
for x, y in it:
    ...
```
and we may invent more complex strategy which would work in all these cases, but for now the simplest strategy is the best that works for most common cases.